### PR TITLE
feat: implement StoreWithFallbacks

### LINKED ISSUE #35: Support creating store with fallbacks
Proposed API:

```go
package credentials

// NewStoreWithFallbacks returns a new store based on the given stores.
// The second and the subsequent stores will be used as fallbacks for the first store.
func NewStoreWithFallbacks(store Store, fallbacks ...Store) Store {
	panic("not implemented") // TODO: Implement
}
```

Design discussion: #18 
Reference: https://github.com/oras-project/oras/blob/main/internal/credential/store.go#L93-L129

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Resolves #35

### CODE REVIEW

File: store_with_fallbacks_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+func TestStoreWithFallbacks(t *testing.T) {
+	ns := &NativeStore{
+		programFunc: testCommandFn,
+	}
+	fs, err := NewFileStore("testdata/valid_auths_config.json")
+	if err != nil {
+		t.Fatal("NewFileStore() error =", err)
+	}
+	sf := NewStoreWithFallbacks(ns, fs)
```
Comment (Wwwsylvia) [inline store_with_fallbacks_test.go:34]: I think using a test store here is enough.

Comment (wangxiaoxuan273) [reply store_with_fallbacks_test.go:34]: Thanks, I totally forgot about test store lol

Comment (wangxiaoxuan273) [reply store_with_fallbacks_test.go:34]: Replaced with test stores.

File: store_with_fallbacks_test.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	"testing"
+
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+func TestStoreWithFallbacks(t *testing.T) {
+	ns := &NativeStore{
+		programFunc: testCommandFn,
+	}
+	fs, err := NewFileStore("testdata/valid_auths_config.json")
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply store_with_fallbacks_test.go:30]: Replaced with test stores.

File: store_with_fallbacks.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	"context"
+
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// StoreWithFallbacks is a store that has multiple fallback stores.
+type StoreWithFallbacks struct {
+	primaryStore   Store
+	fallbackStores []Store
+}
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply store_with_fallbacks.go:28]: Updated accordingly.

**Review by Wwwsylvia — COMMENTED** [review]
We may need better documentations for the behaviors.

File: store_with_fallbacks.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+// StoreWithFallbacks is a store that has multiple fallback stores.
+type StoreWithFallbacks struct {
+	stores []Store
+}
+
+// NewStoreWithFallbacks returns a new store based on the given stores.
+// The second and the subsequent stores will be used as fallbacks for the first store.
+func NewStoreWithFallbacks(store Store, fallbacks ...Store) Store {
```
Comment (Wwwsylvia) [inline store_with_fallbacks.go:31]: I'm still struggling between

```go
NewStoreWithFallbacks(store Store, fallbacks ...Store) Store
```
and 
```go
NewStoreWithFallbacks(stores ...Store) Store
```


Later we will have a docker store which is based on the docker configuration file at the default location. I'm wondering how should we combine it with `StoreWithFallbacks`.

If we go for option 1, the experience would be:
```go
ns := NewNativeStore()
store := NewStoreWithFallbacks(ns, NewStoreFromDocker())
```
or
```go
dockerStore := NewStoreFromDocker()
store := NewStoreWithFallbacks(dockerStore)
```

If we go for option 2, we should have the docker store to be the default store when no store is passed.
So 
```go
dockerStore := NewStoreFromDocker()
store := NewStoreWithFallbacks(dockerStore)
```
is equivalent to

```go
store := NewStoreWithFallbacks()
```

Comment (Wwwsylvia) [reply store_with_fallbacks.go:31]: I think I still prefer option 1.

**Comment by wangxiaoxuan273** [comment]
> We may need better documentations for the behaviors.

Good advice. Added more detailed documentations for `Get,` `Store` and `Delete`.

File: store_with_fallbacks.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+// StoreWithFallbacks is a store that has multiple fallback stores.
+// Please use the NewStoreWithFallbacks to create new instances of
+// StoreWithFallbacks.
+type StoreWithFallbacks struct {
+	stores []Store
+}
+
+// NewStoreWithFallbacks returns a new store based on the given stores.
+// The first store in the array is used as the primary store. The second
```
Comment (Wwwsylvia) [inline store_with_fallbacks.go:32]: > The first store in the array is used as the primary store

This looks a bit strange as this function does not accept an array of stores for the primary store.

File: store_with_fallbacks.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+// and the subsequent stores will be used as fallbacks for the first store.
+func NewStoreWithFallbacks(store Store, fallbacks ...Store) Store {
+	return &StoreWithFallbacks{
+		stores: append([]Store{store}, fallbacks...),
+	}
+}
+
+// Get retrieves credentials from the StoreWithFallbacks for the given server.
+// It searches the array of stores for the credentials of serverAddress
```
Comment (Wwwsylvia) [inline store_with_fallbacks.go:41]: The array of stores is invisible to users.

File: store_with_fallbacks.go
Comment (shizhMSFT) [inline store_with_fallbacks.go:1]: Can we put everything in `store.go` instead of having a new file?

File: store_with_fallbacks.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"context"
+
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// StoreWithFallbacks is a store that has multiple fallback stores.
+// Please use the NewStoreWithFallbacks to create new instances of
+// StoreWithFallbacks.
+type StoreWithFallbacks struct {
```
Comment (shizhMSFT) [inline store_with_fallbacks.go:27]: Do we need to export this struct? Can we make it private?

File: store_with_fallbacks.go (hunk 5 above)
Comment (wangxiaoxuan273) [reply store_with_fallbacks.go:32]: removed the description about the array.

File: store_with_fallbacks.go (hunk 6 above)
Comment (wangxiaoxuan273) [reply store_with_fallbacks.go:41]: removed the description about the array.

File: store_with_fallbacks.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// NewStoreWithFallbacks returns a new store based on the given stores.
+// The first store in the array is used as the primary store. The second
+// and the subsequent stores will be used as fallbacks for the first store.
+func NewStoreWithFallbacks(store Store, fallbacks ...Store) Store {
+	return &StoreWithFallbacks{
+		stores: append([]Store{store}, fallbacks...),
+	}
+}
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply store_with_fallbacks.go:38]: Then it complicates the implementation of Get and Put and Delete. Putting it in the array is fine.

File: store_with_fallbacks.go
Comment (wangxiaoxuan273) [reply store_with_fallbacks.go:1]: moved.

File: store_with_fallbacks.go (hunk 7 above)
Comment (wangxiaoxuan273) [reply store_with_fallbacks.go:27]: Made it private.

File: store_with_fallbacks.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply store_with_fallbacks.go:38]: changed it.

File: store.go [hunk 9]
```diff
@@ -30,3 +30,50 @@ type Store interface {
 	// Delete removes credentials from the store for the given server address.
 	Delete(ctx context.Context, serverAddress string) error
 }
+
+// storeWithFallbacks is a store that has multiple fallback stores.
+// Please use the NewStoreWithFallbacks to create new instances of
+// storeWithFallbacks.
```
Comment (Wwwsylvia) [inline store.go:36]: This looks unnecessary.

File: store.go [hunk 10]
```diff
@@ -30,3 +30,50 @@ type Store interface {
...
+type storeWithFallbacks struct {
+	stores []Store
+}
+
+// NewStoreWithFallbacks returns a new store based on the given stores.
+// The first store is used as the primary store. The second and the
+// subsequent stores will be used as fallbacks for the first store.
+func NewStoreWithFallbacks(store Store, fallbacks ...Store) Store {
+	if fallbacks == nil {
```
Comment (Wwwsylvia) [inline store.go:45]:
```suggestion
	if len(fallbacks) == 0 {
```
`fallbacks` can be not-nil but empty.

File: store.go [hunk 11]
```diff
@@ -30,3 +30,50 @@ type Store interface {
...
+		return store
+	}
+	return &storeWithFallbacks{
+		stores: append([]Store{store}, fallbacks...),
+	}
+}
+
+// Get retrieves credentials from the StoreWithFallbacks for the given server.
+// It searches the primary and the fallback stores for the credentials of serverAddress
+// and returns when it finds the credentials in any of the stores.
```
Comment (Wwwsylvia) [inline store.go:55]: I'm wondering if users can see these comments. We may need to move them to `NewStoreWithFallbacks`.

File: store.go [hunk 12]
```diff
@@ -30,3 +30,50 @@ type Store interface {
...
+		}
+		if cred != auth.EmptyCredential {
+			return cred, nil
+		}
+	}
+	return auth.EmptyCredential, nil
+}
+
+// Put saves credentials into the StoreWithFallbacks. It puts
+// the credentials into the primary store.
```
Comment (Wwwsylvia) [inline store.go:70]: Same here.

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: store.go (hunk 9 above)
Comment (wangxiaoxuan273) [reply store.go:36]: removed.

File: store.go (hunk 10 above)
Comment (wangxiaoxuan273) [reply store.go:45]: changed.

File: store.go (hunk 11 above)
Comment (wangxiaoxuan273) [reply store.go:55]: moved.

File: store.go (hunk 12 above)
Comment (wangxiaoxuan273) [reply store.go:70]: moved.

**Pull request merged**